zdm/iteration.py

```diff
@@ -784,30 +947,28 @@ def calc_DMG_weights(DMEGs, DMGs, sigmaDMGs, dmvals, n_sig=3):
     iweights = []
 
     # Loop through the DMG of each FRB in the survey and determine the weights
-    for i,DMG in enumerate(DMGs):
-        # Get absolute uncertainty in DMG
-        sigmaDMG = DMG * sigmaDMGs
+    for i,DM_ISM in enumerate(DM_ISMs):
+        # Get absolute uncertainty in DM_ISM
+        sigma_ISM_abs = DM_ISM * sigma_ISM
 
         # Determine lower and upper DM values used
-        delta = n_sig*sigmaDMG
-        min = DMEGs[i] - delta
-        # actual DMG can't be less than 0, so if delta is larger than DMG:
-        # DMEG = DM_tot - DMG - DMhalo
-        # DMEG_max = DM_tot - DMhalo = DMEG + DMG
-        if delta > DMG:
-            max = DMEGs[i] + DMG
-        else:
-            max = DMEGs[i] + delta
-
-        # Determine indices of where the DM is between min and max
-        mask = (dmvals > min) * (dmvals < max)
-        idxs = np.where(mask)
+        # From 0 to DM_total
+        DM_total = DMEGs[i] + DM_ISM + DMhalo
+        idxs = np.where(dmvals < DM_total)
 
         # Get weights
-        x = dmvals[idxs]
-        weight = st.norm.pdf(x, loc=DMEGs[i], scale=sigmaDMG)
-        weight = weight / np.sum(weight)    # Re-normalise
-        weights.append(weight)
+        DMGvals = DM_total - dmvals[idxs] # Descending order because dmvals are ascending order
+        ddm = dmvals[1] - dmvals[0]
+
+        pISM = st.norm.pdf(DMGvals, loc=DM_ISM, scale=sigma_ISM_abs) * ddm
+        pHalo = st.norm.pdf(DMGvals, loc=DMhalo, scale=sigma_halo) * ddm
+        
+        pDMG = np.convolve(pISM, pHalo, mode='full')
+        # Set upper limit of DMG = DM_total 
```

Comment:
I probably missed it, but where is the upper limit set?

Reply:
It is done twice. Firstly, DMGvals has an upper limit (from idxs = np.where(dmvals < DM_total)) which places the bounds on pISM and pHalo. Then we take an upper limit on pDMG by taking
pDMG = pDMG[-len(DMGvals):]
DMGvals goes from 0 to DMtot and so just taking those indices corresponds to 0 to DMtot. I hope that makes some sense...